src/main/java/org/apache/solr/benchmarks/indexing/IndexBatchSupplier.java

```diff
@@ -0,0 +1,130 @@
+package org.apache.solr.benchmarks.indexing;
+
+import org.apache.http.client.HttpClient;
+import org.apache.solr.benchmarks.BenchmarksMain;
+import org.apache.solr.benchmarks.beans.IndexBenchmark;
+import org.apache.solr.common.cloud.DocCollection;
+import org.apache.solr.common.cloud.Slice;
+import org.apache.solr.common.util.JsonRecordReader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.StringReader;
+import java.lang.invoke.MethodHandles;
+import java.util.*;
+import java.util.concurrent.*;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Supplier;
+
+public class IndexBatchSupplier implements Supplier<Callable>, AutoCloseable {
+    private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+    private final Future<?> workerFuture;
+    private volatile boolean exit;
+    private IndexBenchmark benchmark;
+    private DocCollection docCollection;
+    private HttpClient httpClient;
+    private Map<String, String> shardVsLeader;
+    private BlockingQueue<UploadDocs> pendingBatches = new LinkedBlockingQueue<>(10); //at most 10 pending batches
+
+    private AtomicLong docsIndexed = new AtomicLong();
+
+    public IndexBatchSupplier(DocReader docReader, IndexBenchmark benchmark, DocCollection docCollection, HttpClient httpClient, Map<String, String> shardVsLeader) {
+        this.benchmark = benchmark;
+        this.docCollection = docCollection;
+
+        this.httpClient = httpClient;
+        this.shardVsLeader = shardVsLeader;
+
+        this.workerFuture = startWorker(docReader);
+    }
+
+    class IdParser implements JsonRecordReader.Handler {
+        String idParsed;
+
+        @Override
+        public void handle(Map<String, Object> record, String path) {
+            idParsed = record.get(benchmark.idField) instanceof String ?
```

Comment:
if idField is missing you may get an NPE

Reply:
AFAIK, the bean should be `idField` defeault as "id". This Parser is refactoring of existing code from https://github.com/fullstorydev/solr-bench/pull/50/files/07565fdeb86670e2e32057cb893d0605a836a14e#diff-c443b311a116a519dcdbf35939db1b752571af41e44a38f37368dc4770092ab1L318 so it should be rather safe 😊